File: native/engine-chromium-desktop/src/handlers/renderer_handler.cpp
#include "renderer_handler.h"

namespace mollotov {

RendererHandler::RendererHandler(DesktopHandlerRuntime runtime) : runtime_(std::move(runtime)) {}

void RendererHandler::Register(DesktopRouter& router) const {
  router.Register("get-renderer", [this](const nlohmann::json&) { return GetRenderer(); });
  router.Register("set-renderer", [this](const nlohmann::json&) { return SetRenderer(); });
}

nlohmann::json RendererHandler::GetRenderer() const {
  if (runtime_.renderer_supplier) {
    return runtime_.renderer_supplier();
  }
  return SuccessResponse({{"current", runtime_.engine_name}, {"available", {"chromium"}}});
}

nlohmann::json RendererHandler::SetRenderer() const {
  return Unsupported("set-renderer");
}

}  // namespace mollotov

File: apps/macos/Kelpie/Handlers/RendererHandler.swift
import Foundation

/// Handles set-renderer and get-renderer endpoints.
/// Triggers cookie migration and engine swap.
struct RendererHandler {
    let context: HandlerContext
    let rendererState: RendererState
    let onSwitch: (RendererState.Engine) async -> Void

    func register(on router: Router) {
        router.register("set-renderer") { body in await setRenderer(body) }
        router.register("get-renderer") { _ in await getRenderer() }
    }

    @MainActor
    private func setRenderer(_ body: [String: Any]) async -> [String: Any] {
        guard let engineStr = body["engine"] as? String else {
            return errorResponse(code: "MISSING_PARAM", message: "engine is required (webkit|chromium)")
        }
        guard let engine = RendererState.Engine(rawValue: engineStr) else {
            return errorResponse(code: "INVALID_PARAM", message: "engine must be webkit or chromium")
        }
        if engine == rendererState.activeEngine {
            return successResponse(["engine": engine.rawValue, "changed": false])
        }

        await onSwitch(engine)

        return successResponse(["engine": engine.rawValue, "changed": true])
    }

    @MainActor
    private func getRenderer() async -> [String: Any] {
        successResponse([
            "engine": rendererState.activeEngine.rawValue,
            "available": RendererState.Engine.allCases.map(\.rawValue)
        ])
    }
}
